### PR TITLE
Write the 1.4 release notes

### PR DESCRIPTION
**Pull request by maelvls**
| Preview | no preview since the Netlify previews are only done on `master`, not `release-next` |
|-|-|

Sorry for the delay, I hope that the review of this PR won't affect tomorrow's 1.4 release.

cc @irbekrm @wallrj 

Command used to generate the notes:

```sh
export RELEASE_VERSION=1.4.0
export BRANCH=release-1.4
export END_REV=master
export START_REV=v1.3.0
export GITHUB_TOKEN=$(lpass show github.com -p)
release-notes --debug --repo-path cert-manager \
  --org jetstack --repo cert-manager \
  --required-author "jetstack-bot" \
  --output release-notes.md
```

### CODE REVIEW

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [inline content/en/docs/release-notes/release-notes-1.4.md:25]: I moved this to the "Bug Fixes" section (first entry) since I think it is now a breaking change anymore... or is it still?

@irbekrm

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [inline content/en/docs/release-notes/release-notes-1.4.md:164]: this link fails since it only exists in next-docs right now. When release-next is merged into master, the issue will go away. I don't know how to work around Prow regarding that though

Should I use the link 

```md
[the documentation](/next-docs/usage/kube-csr)
```

instead?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md:25]: It does change the renewal behaviour if your certs have a duration of more than 90 days and `renewBefore` has not been set. Previously in that case the certs would have been renewed 30 days before expiry whereas now it will be 2/3 through the duration. Do you think that's worth being added as a change?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md:164]: > When release-next is merged into master

Since we are releasing today, do you think we can just somehow co-ordinate the merging so that we merge release-next into master before merging this PR?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [inline content/en/docs/release-notes/release-notes-1.4.md:276]: Suggestion: 'who has been accepted'

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [inline content/en/docs/release-notes/release-notes-1.4.md:270]: Nit: I think this may have been for the 1.3 release?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [inline content/en/docs/release-notes/release-notes-1.4.md:247]: This was done so it's possible to upgrade from `v1.2.x` to  `v1.3.x`. I think this might not be relevant for `v1.4`?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [inline content/en/docs/release-notes/release-notes-1.4.md:278]: I think Tim will work on install command for kubectl cert-manager plugin?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md:164]: Also looking at the error message, might it just be the trailing slash?:

```
docs/release-notes/release-notes-1.4/index.html
  target is a directory, href lacks trailing slash --- docs/release-notes/release-notes-1.4/index.html --> ../../usage/kube-csr
```

**Review by irbekrm — CHANGES_REQUESTED** [review]
Looks great to me, thanks for taking the time to write all this up!

I have added a couple of small suggestions.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:243]: The revert took place in https://github.com/jetstack/cert-manager/commit/79d8d9cb7b5186f0fcffe2e24e1198656737f26d, so I assumed that this PR was meant to be merged as part of the 1.4 release?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:220]: Should I rephrase as

```suggestion
- The upgrade from 1.2 to 1.3 with Helm and `--set installCRDs=true` is
```

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:252]: I thought we had submitted a new proposal, it [looks like not](https://github.com/cncf/mentoring/pulls?q=cert-manager)... Sorry, I will rephrase that and say that he is working on the CLI stuff.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:45]: Can we be more specific about which users are affected by this? 

> If you are upgrading cert-manager on a cluster which has previously had cert-manager < v1.0.0 you MUST take steps to upgrade any cert-manager custom resources which were created by you or by cert-manager < v1.0.0.
> Such custom resources may still be stored in Etcd as v1alpha2 resources and when that version of the API is removed in cert-manager 1.6, they will become unreadable. 
> To work around this, you need to force them to be converted by the  cert-manager conversion webhook.
> And the simplest way to do that is to simply perform a no-change update on all cert-manager customer resources. 
> This is explained in more detail in the [Upgrading existing cert-manager resources] page.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Should I still apply this change? The sentence is now (a5a3b27):

Tim [has been
accepted](https://summerofcode.withgoogle.com/projects/#4841881566969856) for
the 2021 Google Summer of Code to work on the `kubectl cert-manager install`
command.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md]: I can't stand links like `now` and `docs` and `here and `this`. 
I suggest not linking to the PR.
```suggestion
The Helm chart now follows
the [Kyverno restricted Pod security policy](https://kyverno.io/policies/pod-security/#restricted).
```

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:79]: I suggest removing the `now` PR link.
And if you want to add PR links, do it at the end of each note as `Implemented in cert-manager PR#3865` (or something like that)

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md]: > I have tested v1.3.0 -> latest master (so with this PR) upgrade on kind clusters with k8s v1.20 and v1.19 (with helm)

The goal of the PR was to fix a bug where users couldn't upgrade _to_ `v1.3.x`. The other comments are mostly for edge cases. We also didn't merge this PR into `v1.3.0`.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md]: I think it _may_ be this issue that is related to his proposal https://github.com/jetstack/cert-manager/issues/3952

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md:244]: You're right. We _wanted_ to get it into v1.3 release 😄

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:84]:
```suggestion
⛔️  You may need to adjust systems that consume the `ca.crt` from Secrets managed by cert-manager with the CA issuer.
```

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:25]: Yes, I think that's a notable (and maybe breaking) change.

I moved the bug fix from under "Bug Fixes" to "Breaking Changes" in 119faf9 with a way to keep the old behavior by setting renewBefore to `720h`.

What do you think?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:150]: This seems like too much documentation to include in the release-notes. It should be in the kube-csr docs which we link to.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:167]: "start of the chain" is ambiguous. 
If we are going to explain it, it should be in the docs for the Vault issuer.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:175]: I like that you're crediting Josh's code here, but it seems inappropriate for the release notes.
After all, you might argue that we should have been doing this from the start and we should be applying this cert chain parsing to the CertificateRequest CA information returned by all the issuers. 

And the fact that we're saying here that the CA may contain an intermediate just adds to the general confusion around `ca.crt` and contradicts the CA bug fix described above where we say that the `ca.crt` always contains a root crt.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md]: Again, I suggest updating the webhook docs and linking to that.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (wallrj) [inline content/en/docs/release-notes/release-notes-1.4.md:256]: I know I'm too straight laced, but I just don't think this is appropriate for the release notes. It would be appropriate for a blog post.
Mentioning a not-merged feature in the release notes is just going to confuse people I think.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Lol once again I did not read the error message and assumed it was a broken link. Thank you for taking a look at the Prow build logs 🙏

**Review by wallrj — CHANGES_REQUESTED** [review]
Thanks @maelvls 

Most of my comments are below are subjective / advisory only. Generally:

* I think the details of each feature / fix should be moved out of release notes and into the appropriate section of the documentation.
* Hyperlinks should contain the target document titles 
* Honorable mentions should be moved to a blog post.

My 2c.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Wait, so this PR is only meant to fix e.g. `v1.2.0` -> `v1.3.1` but we did not backport it to `release-1.3`?

Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: I shall then create a new `1.3.1` entry in the 1.3 release notes page 😅

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: I agree, it kind of duplicates the content of usage/kube-csr. I removed most of it in b0085737aa3e2803d2a18ebe871165d5b2c1572b.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: I think it is not worth detailing this, so I removed the mention to "start of the chain" in a3d6a65.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: I removed this part in 827ec9e

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md]: This looks great to me!

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md]: I think this is correct - @wallrj could confirm that?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (SgtCoDFish) [inline content/en/docs/release-notes/release-notes-1.4.md:47]: suggestion: with this being in a warning which will draw peoples' attention, we should be really clear that this isn't an _immediate_ call to action.

```suggestion
⛔️  If you have a cert-manager installation that is using or has previously used
these deprecated APIs, you may need to upgrade your cert-manager custom
resources and CRDs before the release of cert-manager version 1.6.

See the ["remove deprecated APIs" docs](../../installation/upgrading/remove-deprecated-apis/#upgrading-existing-cert-manager-resources)
 for more detailed upgrade instructions.
```

**Review by SgtCoDFish — COMMENTED** [review]
A few little bits from me!

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: After [a discussion](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1623770041428700) with Josh, I rephrased the paragraph in a1744c9. Can you take a look?

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [inline content/en/docs/release-notes/release-notes-1.4.md:95]: @erikgb @JoshVanL @wallrj does this paragraph make sense?

Context: [this Slack thread](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1623770041428700)

Comment (JoshVanL) [reply content/en/docs/release-notes/release-notes-1.4.md:95]: Makes sense to me :+1:

Comment (erikgb) [reply content/en/docs/release-notes/release-notes-1.4.md:95]: Makes sense to me too, but this is a definately bit "complex". The `ca.crt` will contain the "most" root CA it is aware of. 😄

Comment (erikgb) [reply content/en/docs/release-notes/release-notes-1.4.md:95]: I mean: The self-signed CA, aka root-CA is not really required to issue certificates.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Thank you for this great suggestion, I added it verbatim in 647c958.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Richard suggested a new phrasing in 647c958 which removes the feeling of "it should be done immediately". Let me know if that's still feeling too much.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:104]: Thanks Erik!! Your phrasing is clearer. I went with:

> The CA issuer now produces a `ca.crt` that contains the "most" root CA that cert-manager is aware of. `ca.crt` may thus not be the actual self-signed root CA, since cert-manager may not be aware of it.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md:164]: Thanks for the suggestion, I was not aware of "example IPv4 blocks"! 👍

**Comment by maelvls** [comment]
Should I also add a section "Cleanup or Flakes"? Running release-notes, there are quite a lot of entries:

- Cert-manager controller now uses ConfigMapsLeasesResourceLock for leader election. ([#4016](https://github.com/jetstack/cert-manager/pull/4016), [@irbekrm](https://github.com/irbekrm))
- Change Venafi Issuer to populate CertificateRequest.Status.CA with the root most certificate that was returned from signing. ([#3983](https://github.com/jetstack/cert-manager/pull/3983), [@JoshVanL](https://github.com/JoshVanL))
- Deprecate Issuer.spec.acme.externalAccountBinding.keyAlgorithm field. EAB MAC algorithm is now hardcoded to HS256. ([#3877](https://github.com/jetstack/cert-manager/pull/3877), [@irbekrm](https://github.com/irbekrm))
- Deprecates UsageContentCommittment ([#3860](https://github.com/jetstack/cert-manager/pull/3860), [@jsoref](https://github.com/jsoref))
- Deprecates cert-manager.io/v1alpha2, cert-manager.io/v1alpha3, cert-manager.io/v1beta1, acme.cert-manager.io/v1alpha2, acme.cert-manager.io/v1alpha3, acme.cert-manager.io/v1beta1 APIs. These APIs will be removed in cert-manager v1.6 ([#4021](https://github.com/jetstack/cert-manager/pull/4021), [@irbekrm](https://github.com/irbekrm))
- Optimistic locking messages (the object has been modified) are now logged at the Info level instead of the Error level, as cert-manager controllers will automatically retry until successful. ([#3794](https://github.com/jetstack/cert-manager/pull/3794), [@JoshVanL](https://github.com/JoshVanL))
- Panic when failing to register schemes during initialization for pkg/webhook/server
  Various static analysis fixes across many files including removing unused or redundant code ([#4037](https://github.com/jetstack/cert-manager/pull/4037), [@SgtCoDFish](https://github.com/SgtCoDFish))
- REMOVE ME WHEN RELEASING! This PR reverts a feature with a release-note,
  so ensure the final generated release notes don't contain a reference to the removed feature. ([#3988](https://github.com/jetstack/cert-manager/pull/3988), [@jakexks](https://github.com/jakexks))
- Removes legacy util functions for issuer generation from test/e2e/util/util.go. Use functions in test/unit/gen/issuer.go instead. ([#3873](https://github.com/jetstack/cert-manager/pull/3873), [@irbekrm](https://github.com/irbekrm))
- Testing: Adds Kubernetes CertificateSigningRequest CA Issuer E2E tests. ([#4081](https://github.com/jetstack/cert-manager/pull/4081), [@JoshVanL](https://github.com/JoshVanL))
- The ACME issuer now constructs a certificate chain after signing, and populates the CertificateRequest.Status.CA with the root most certificate if available. ([#3984](https://github.com/jetstack/cert-manager/pull/3984), [@JoshVanL](https://github.com/JoshVanL))
- Update default development version of Kubernetes in kind from 1.17 to 1.20 ([#4015](https://github.com/jetstack/cert-manager/pull/4015), [@SgtCoDFish](https://github.com/SgtCoDFish))
- Updated details of FindZoneByFqdn error message when an unexpected DNS response code is received. ([#3906](https://github.com/jetstack/cert-manager/pull/3906), [@clatour](https://github.com/clatour))
- Updates Kubernetes libaries to v1.21.0 ([#3926](https://github.com/jetstack/cert-manager/pull/3926), [@tamalsaha](https://github.com/tamalsaha))
- Updates distroless/static base image to latest version as of 2021-05-20 ([#4039](https://github.com/jetstack/cert-manager/pull/4039), [@SgtCoDFish](https://github.com/SgtCoDFish))
- Validating webhook returns a warning if the legacy ACME issuer EAB key algorithm is set. ([#3936](https://github.com/jetstack/cert-manager/pull/3936), [@irbekrm](https://github.com/irbekrm))

**Review by wallrj — APPROVED** [review]
I think this reads nicely.

/lgtm
/approve
/hold

Holding in case there is still work to be done. 
@maelvls unhold if you're now happy with it.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (irbekrm) [inline content/en/docs/release-notes/release-notes-1.4.md:54]:
```suggestion
⛔️  If you are upgrading cert-manager on a cluster which has previously had
cert-manager < `v1.0.0`, you will need to ensure that all cert-manager custom resources
are stored in etcd at `v1` version and that cert-manger CRDs do not reference the deprecated APIs
**by the time you upgrade to `v1.6`**.

This is explained in more detail in the [Upgrading existing cert-manager
resources][upgrade-resources] page.
```

Comment (irbekrm) [reply content/en/docs/release-notes/release-notes-1.4.md:54]: Let me know if this makes sense ^. I think it might make sense to give this a bit less urgency- users only need to do this by the time they upgrade to v1.6 not now, so it might be confusing.

**Comment by maelvls** [comment]
Done! Could you take another look @wallrj? I re-added the links, let me know if that fits better. I also added the "bug fixes" section.

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: I'll click "Resolve conversation" for now, feel free to re-open if you think this "Honorable mentions" should not be there @wallrj

File: content/en/docs/release-notes/release-notes-1.4.md
Comment (maelvls) [reply content/en/docs/release-notes/release-notes-1.4.md]: Thanks! I also changed the title of the PR (issuer -> DNS01 solver)

**Review by wallrj — APPROVED** [review]
/lgtm
/unhold

**Pull request merged**
